jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRabitRegressionSuite.scala

```diff
@@ -0,0 +1,121 @@
+      "objective" -> "binary:logistic", "num_round" -> 5, "num_workers" -> numWorkers)
+    ).fit(training)
+    val prediction1 = model1.transform(testDF).select("prediction").collect()
+
+    val model2 = new XGBoostClassifier(Map("eta" -> "1", "max_depth" -> "2", "verbosity" -> "1",
+      "objective" -> "binary:logistic", "num_round" -> 5, "num_workers" -> numWorkers,
+      "rabit_ring_reduce_threshold" -> 1))
+      .fit(training)
+
+    assert(Rabit.rabitEnvs.asScala.size > 3)
+    Rabit.rabitEnvs.asScala.foreach( item => {
+      if (item._1.toString == "rabit_reduce_ring_mincount") assert(item._2 == "1")
+    })
+
+    val prediction2 = model2.transform(testDF).select("prediction").collect()
+    // check parity w/o rabit cache
+    prediction1.zip(prediction2).foreach { case (Row(p1: Double), Row(p2: Double)) =>
+      assert(p1 == p2)
+    }
+  }
+
+  test("test regression prediction parity w/o ring reduce") {
+    val training = buildDataFrame(Regression.train)
+    val testDM = new DMatrix(Regression.test.iterator, null)
+    val testDF = buildDataFrame(Classification.test)
+
+    val model1 = new XGBoostRegressor(Map("eta" -> "1", "max_depth" -> "2", "verbosity" -> "1",
+      "objective" -> "reg:squarederror", "num_round" -> 5, "num_workers" -> numWorkers)
+    ).fit(training)
+    val prediction1 = model1.transform(testDF).select("prediction").collect()
+
+    val model2 = new XGBoostRegressor(Map("eta" -> "1", "max_depth" -> "2", "verbosity" -> "1",
+      "objective" -> "reg:squarederror", "num_round" -> 5, "num_workers" -> numWorkers,
+      "rabit_ring_reduce_threshold" -> 1)).fit(training)
+    assert(Rabit.rabitEnvs.asScala.size > 3)
+    Rabit.rabitEnvs.asScala.foreach( item => {
+      if (item._1.toString == "rabit_reduce_ring_mincount") assert(item._2 == "1")
+    })
+    // check the equality of single instance prediction
+    val prediction2 = model2.transform(testDF).select("prediction").collect()
+    // check parity w/o rabit cache
+    prediction1.zip(prediction2).foreach { case (Row(p1: Double), Row(p2: Double)) =>
+      assert(math.abs(p1 - p2) < predictionErrorMin)
+    }
+  }
+
+  test("test rabit timeout failure graceful shutdown") {
+    // disable task fail listener
+    TaskFailedListener.killerStarted = true
+
+    val training = buildDataFrame(Classification.train)
+    // mock rank 0 failure during 4th allreduce synchronization
+    Rabit.mockList = Array("0,8,0,0").toList.asJava
+    intercept[SparkException] {
+      new XGBoostClassifier(Map(
+        "eta" -> "0.1",
+        "max_depth" -> "10",
+        "verbosity" -> "1",
+        "objective" -> "binary:logistic",
+        "num_round" -> 5,
+        "num_workers" -> numWorkers,
+        "rabit_timeout" -> 0))
+        .fit(training)
+    }
+  }
+
+  test("test graceful handle task multiple falures") {
+    // disable task fail listener
+    TaskFailedListener.killerStarted = true
```

Comment:
why we need to do this since we are not testing recovery functionality?

Reply:
https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/ParameterSuite.scala 

you don't need to capture an exception, 

based on the current behavior of xgb-spark, SparkContext is shutdown properly instead of terminating JVM process so that the whole unit test is terminated is the result of "a graceful shutdown"

check https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/ParameterSuite.scala

Reply:
removed multiple failure as spark context will stop after first failure.